contributors/design-proposals/storage/online-grow-volume-size.md

```diff
@@ -0,0 +1,284 @@
+
+## Volume Plugin Matrix
+
+<table border="2" align="center">
+<tr>
+<th>Volume plugin</th>
+<th>Supports Resize</th>
+<th>Requires File system Resize</th>
+<th>Support online file system resize</th>
+<th>Support in 1.9 release</th>
+</tr>
+<tr><td>EBS</td><td>Yes</td><td>Yes</td><td>Yes</td><td>Yes</td></tr>
+<tr><td>GCE PD</td><td>Yes</td><td>Yes</td><td>Yes</td><td>Yes</td></tr>
+<tr><td>GlusterFS</td><td>Yes</td><td>No</td><td>No</td><td>Yes</td></tr>
+<tr><td>Cinder</td><td>Yes</td><td>Yes</td><td>No</td><td>Yes</td></tr>
+<tr><td>Ceph RBD</td><td>Yes</td><td>Yes</td><td>Yes</td><td>No</td></tr>
+<tr><td>Vsphere</td><td>Yes</td><td>Yes</td><td>Yes</td><td>No</td></tr>
+<tr><td>Host Path</td><td>No</td><td>No</td><td>No</td><td>No</td></tr>
+<tr><td>Azure Disk</td><td>Yes</td><td>Yes</td><td>Yes</td><td>No</td></tr>
+<tr><td>Azure File</td><td>No</td><td>No</td><td>No</td><td>No</td></tr>
+<tr><td>Cephfs</td><td>No</td><td>No</td><td>No</td><td>No</td></tr>
+<tr><td>NFS</td><td>No</td><td>No</td><td>No</td><td>No</td></tr>
+<tr><td>Flex</td><td>Yes</td><td>Maybe</td><td>Maybe</td><td>No</td></tr>
+<tr><td>LocalStorage</td><td>Yes</td><td>Yes</td><td>No</td><td>No</td></tr>
+</table>
+
+## Implementation Design
+
+### Overview
+
+The key point of online file system resizing is how kubelet to discover which  PVCs need file system resizing. We achieve this goal by adding a `volumeFSResizingAnnotation` annotation to pod. In detail, if a volume received an online file system resizing request, the annotation will be added to the pods using this volume, so kubelet can watch for pods running on it to discover this annotation, and then start the online file system resizing operation for according volumes.
+
+The `volumeFSResizingAnnotation` annotation's key is `volumes.kubernetes.io/fs-resizing-{volumeName}`, and set it's value to `yes` if online file system resizing if required. We can use `volume.Spec.Name()` as the `volumeName` to guarantee that:
+
+* Annotation's key is a qualified name
+* `volumeName` is unique for a single pod
+
+### Prerequisite
+
+* `sig-api-machinery` need to allow pod's annotation update from kubelet. 
+* `sig-api-machinery` need to allow pod's get and annotation update from expand-controller.
+
+### Controller Manager Resize
+
+#### Fetch All Pods Using a PVC
+
+We need to know which pods are using a volume with specified name, but `ExpandController` doesn't cache this relationship. To achieve this goal, We can share the `DesiredStateOfWorld` with `AttachDetachController`. In detail, reflector the `pkg/controller/volume/attachdetach/cache` package to `pkg/controller/volume/cache`, and add a `GetPodsInVolume` method to `DesiredStateOfWorld`. Expand controller use this method to fetch pods using the volume with specified name.
+
+The `GetPodsInVolume` method will look like this:
+
+```go
+func (dsw *desiredStateOfWorld) GetPodsInVolume(volumeName v1.UniqueVolumeName) []*v1.Pod {
```

Comment:
We should really use ActualStateOfWorld rather than dsow here.

Reply:
Yeah, you are right. I think we should get the pod-volume-relationship(which pod use which volume) from `DesiredStateOfWorld` as `ActualStateOfWorld` doesn't cache this relationship, and get the volume-node-relationship(which volume mount to which node) from `ActualStateOfWorld`. The current proposal misses the latter one, I will update the proposal later.

Reply:
I added a detailed explanation for this subject to the proposal.